test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala

```diff
@@ -0,0 +1,365 @@
+
+import java.util
+
+import scala.language.implicitConversions
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+import org.scalajs.jasminetest.JasmineTest
+
+import scala.scalajs.runtime.UndefinedBehaviorError
+
+import java.{util => ju}
+import ju.TreeSet
+import ju.Comparator
+
+
+object TreeSetTest extends TreeSetTest(new TreeSetFactory)
+object TreeSetWithNullTest extends TreeSetTest(new TreeSetWithNullFactory)
+
+class TreeSetTest[F <: TreeSetFactory](treeSetTestFactory: F)
+    extends AbstractSetTest[F](treeSetTestFactory)
+    with SortedSetTest
+    with NavigableSetTest {
+
+  override def testApi(): Unit = {
+    super.testApi()
+    testSortedSetApi(treeSetTestFactory)
+    testNavigableSetApi(treeSetTestFactory)
+
+    it("should store and remove ordered integers") {
+      val ts = treeSetTestFactory.empty[Int]
+
+      expect(ts.size()).toEqual(0)
+      expect(ts.add(222)).toBeTruthy
+      expect(ts.size()).toEqual(1)
+      expect(ts.add(111)).toBeTruthy
+      expect(ts.size()).toEqual(2)
+      expect(ts.first).toEqual(111)
+      expect(ts.remove(111)).toBeTruthy
+
+      expect(ts.size()).toEqual(1)
+      expect(ts.first).toEqual(222)
+
+      expect(ts.remove(222)).toBeTruthy
+      expect(ts.size()).toEqual(0)
+      expect(ts.isEmpty).toBeTruthy
+      expect(ts.remove(333)).toBeFalsy
+      expect {
+        try {
+          ts.first
+          false
+        } catch {
+          case _: NoSuchElementException => true
+          case _: Throwable => false
+        }
+      }.toBeTruthy
+
+      if (treeSetTestFactory.allowsNullElement) {
+        expect(ts.add(null.asInstanceOf[Int])).toBeTruthy
+        expect(ts.contains(null)).toBeTruthy
+        expect(ts.remove(null)).toBeTruthy
+        expect(ts.contains(null)).toBeFalsy
+      }
+    }
+
+    it("should store and remove ordered strings") {
+      val ts = treeSetTestFactory.empty[String]
+
+      expect(ts.size()).toEqual(0)
+      expect(ts.add("222")).toBeTruthy
+      expect(ts.size()).toEqual(1)
+      expect(ts.add("111")).toBeTruthy
+      expect(ts.size()).toEqual(2)
+      expect(ts.first).toEqual("111")
+      expect(ts.remove("111")).toBeTruthy
+
+      expect(ts.size()).toEqual(1)
+      expect(ts.first).toEqual("222")
+
+      expect(ts.remove("222")).toBeTruthy
+      expect(ts.size()).toEqual(0)
+      expect(ts.remove("333")).toBeFalsy
+      expect(ts.isEmpty).toBeTruthy
+
+      if (treeSetTestFactory.allowsNullElement) {
+        expect(ts.add(null)).toBeTruthy
+        expect(ts.contains(null)).toBeTruthy
+        expect(ts.remove(null)).toBeTruthy
+        expect(ts.contains(null)).toBeFalsy
+      }
+    }
+
+    case class TestObj(num: Int)
+
+    when("compliant-asinstanceofs").
+    it("should throw exception on non comparable objects") {
+      val ts = treeSetTestFactory.empty[TestObj]
+
+      expect(ts.size()).toEqual(0)
+      expectThrows[ClassCastException](ts.add(TestObj(111)))
```

Comment:
It seams that the `expectsThrows` is letting the exception pass through. 

```
[error]  x should throw exception on non comparable objects
org.scalajs.jasminetest.JasmineTestException: java.lang.ClassCastException: TestObj(222) is not an instance of java.lang.Comparable
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/library/src/main/scala/scala/scalajs/runtime/StackTrace.scala:39)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/library/src/main/scala/scala/scalajs/runtime/StackTrace.scala:33)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalanglib/src/main/scala/java/lang/Throwables.scala:24)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalanglib/src/main/scala/java/lang/Throwables.scala:12)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalanglib/src/main/scala/java/lang/Throwables.scala:343)
	at <jscode>.(scalajsenv.js:171)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalanglib/src/main/scala/java/lang/Comparable.scala:3)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala:362)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalib/src/main/scala/java/util/TreeSet.scala:35)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/javalib/src/main/scala/java/util/TreeSet.scala:112)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala:109)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala:105)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala:105)
	at <jscode>.(file:/localhome/jenkins/a/workspace/scalajs-task-worker/jasmine-test-framework/src/main/scala/org/scalajs/jasminetest/JasmineTest.scala:21)
```
The issue seams to be with the `excpectThrows` and not with the `TreeSet` implementation. Maybe @sjrd has some insight of why in this case the `excpectThrows` is not catching the exception.

Maybe try to de-alias `ClassCastException` manually by replacing it with `java.lang.ClassCastException`.